atomic bootloader update for SoCs with redundant copy support (#918)
rauc/rauc: docs/advanced.rst

  region-start=1M
  region-size=64M

.. _sec-raw-partition-fallback:

Update Raw Boot Partition with Fallback
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Some SoCs (like the Rockchip RK3568) contain a fixed ROM code that searches the
possible boot media for valid images in a defined order.

> Usually this happens at two or more fixed addresses.

Reviewer: ```suggestion
Usually this happens at two or more fixed addresses ("locations").
```

Special headers and/or magic numbers are typically used to determine if there

> is a valid image at such an address.

Author: Yes, perhaps. To be honest, I'm not familiar with the exact behavior of many ROM code loaders.
However, since that could likely be the case, I'll mention that as well.

This behavior can be used to implement a failsafe update of a bootloader.

> To do this, the bootloader, which starts with the required header, is
> installed in two places that the ROM code searches for possible boot images.
> For example, usually only the code at the first address is used, and the second
> is redundant and unused.

Reviewer: ```suggestion
To do this, the bootloader, which starts with the required header, is
installed in two locations that the ROM code searches for possible boot images.
For example, usually only the code in the first location is used, and the second
is redundant and ignored.
```

> During an update, the unused code is updated first, and then the area that is
> normally started from. It is important that the corresponding header is always

Reviewer: ```suggestion
During an update, the currently unused location is updated first, and followed by the other
location (which as likely just booted from).
It is important that the corresponding header is always
```

deleted first, then the boot image is written and the header is only written
last.
This ensures that there is always a valid image on either address and that
half-written images are not attempted to boot from.

The required slot type is ``boot-raw-fallback``.
The device to be specified is expected to be the underlying block device.

> The boot partitions are derived by the definition of the values
> ``region-start`` and ``region-size``.

Reviewer: Using _partitions_ here could be confusing.
```suggestion
The location of each copy in the boot region is derived from the values
``region-start`` and ``region-size``.
```

Both values have to be set in integer decimal bytes and can be post-fixed with
K/M/G/T.

A ``system.conf`` section could look like this:

.. code-block:: cfg

  [slot.bootloader.0]
  device=/dev/mmcblk0
  type=boot-raw-fallback
  region-start=32k
  region-size=4M

It defines a region starting at ``0x8000`` with a size of ``4M``.
This region will be split up into two region halves of equal size by RAUC
internally.
This results in two regions, one starting at ``0x8000`` and one at
``0x208000``, both with a size of ``2M``.
The first region defines the area where the normally used bootloader should be
located and the second defines the location of the fallback area.

Since the implementation makes certain assumptions, it is important that the
SoC ROM code tries to boot from the first region first.

> Note that under normal circumstances the update will work fine even if the two
> regions are swapped.
> However, for the update to actually be failsafe, the areas must match the
> actual search order of the SoC ROM code.

Reviewer: ```suggestion
Note that under most circumstances the update will appear to work fine even if the two
regions are swapped.
However, for the update to actually be failsafe, the areas must be searched in order by SoC ROM code.
```

We should also have a note that the header size is currently fixed to 512 bytes.

Author: Instead of checking and throwing an error, it might be better to change img_to_boot_raw_fallback_handler() to use the actual sector size as the header length instead of the fixed 512 bytes.
What do you think?

Author: Now the physical sector size is used, which is also mentioned in a small note in the documentation.

Author: Since I've switched back to the fixed head size, I've also added a note on the documentation.
